### PR TITLE
Week 18 - Sofia AW

### PR DESCRIPTION
**Pull request by Sofarz**
Finished project.

https://sofias-mongo-db-api.herokuapp.com/

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,39 +1,172 @@
...
+
+// Middleware - since it is mounted before the route, this middlewear will execute first. 
+// It checks if the database is in a good state or not. If so, invoke next() and continue on to show our data, if not show error message.
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({ error: 'Service unavailable' })
+  }
+})
```
Comment (liaback-lowstett) [inline server.js:85]: Nice!

File: server.js [hunk 2]
```diff
@@ -1,39 +1,172 @@
...
+  res.send(listEndpoints(app))
 })
 
-// Start the server
-app.listen(port, () => {
-  // eslint-disable-next-line
-  console.log(`Server running on http://localhost:${port}`)
+// Endpoint to get all books 
+app.get('/books', async (req, res) => {
+  const books = await Book.find().populate('authors')
```
Comment (liaback-lowstett) [inline server.js:94]: Smart. will have a look att this .populate() method for sure

File: server.js [hunk 3]
```diff
@@ -1,39 +1,172 @@
...
+
+  try {
+    const singleBook = await Book.findById(bookId).populate('authors')
+    if (singleBook) {
+      res.json({ length: singleBook.length, data: singleBook }) 
+    } else {
+      res.status(404).json({ error: 'No book with that id was found!'})
+    }
+  } catch (err) {
+    res.status(400).json({ error: 'Invalid request'})
```
Comment (liaback-lowstett) [inline server.js:110]: nice work with the  status code and the error messages

File: README.md [hunk 4]
```diff
@@ -1,13 +1,41 @@
...
+- https://sofias-mongo-db-api.herokuapp.com/authors 
+  Endpoint to get all authors 
+
+- https://sofias-mongo-db-api.herokuapp.com/authors/:authorId
+  Endpoint to get author based on id path param
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+https://sofias-mongo-db-api.herokuapp.com/
```
Comment (liaback-lowstett) [inline README.md:41]: Looks great! All the endpoints work just the way they should.

**Review by liaback-lowstett — COMMENTED** [review]
Clean and neat project! I don't really have anything to mark on. Good job!

File: README.md [hunk 5]
```diff
@@ -1,13 +1,41 @@
...
+  Endpoint to get author by path param id for single book
+
+- https://sofias-mongo-db-api.herokuapp.com/books/search
+  Endpoint with search path and query param for title
+
+- https://sofias-mongo-db-api.herokuapp.com/authors 
+  Endpoint to get all authors 
+
+- https://sofias-mongo-db-api.herokuapp.com/authors/:authorId
+  Endpoint to get author based on id path param
```
Comment (KarolCislowski) [inline README.md:38]: Nice documentation 👍🏻

File: server.js [hunk 6]
```diff
@@ -1,39 +1,172 @@
...
+  authors: String
+})
+
+// The seedDB is going to re-run everytime we start the server, if we don't want that
+// to happen and want to be able to choose when that should happen, we can wrap it
+// in an environment variable
+if (process.env.RESET_DATABASE) {
+  const seedDB = async () => {
+    await Book.deleteMany()
+    await Author.deleteMany()
```
Comment (KarolCislowski) [inline server.js:49]: Data nicely split into 2 separate sets 🤓

File: server.js [hunk 7]
```diff
@@ -1,39 +1,172 @@
...
-  // eslint-disable-next-line
-  console.log(`Server running on http://localhost:${port}`)
+// Endpoint to get all books 
+app.get('/books', async (req, res) => {
+  const books = await Book.find().populate('authors')
+  res.json({ length: books.length, data: books })
 })
+
+// Endpoint to get single book by id path param
+app.get('/books/book/:bookId', async (req, res) => {
```
Comment (KarolCislowski) [inline server.js:99]: `/books/:bookId` - id is a primary key so you should add it immediately after set name in your endpoint

File: server.js [hunk 8]
```diff
@@ -1,39 +1,172 @@
...
+      title: {
+        $regex: new RegExp(title, 'i')
+      }
+    }).populate('authors')
+    res.json({ length: queriedBook.length, data: queriedBook })
+  } else {
+    const books = await Book.find().populate('authors')
+    res.json({ length: books.length, data: books })
+  }
+})
```
Comment (KarolCislowski) [inline server.js:146]: nice filtering 👍🏻

**Review by KarolCislowski — APPROVED** [review]
Great API 👍🏻

**Pull request closed without merging**